<a href="https://colab.research.google.com/github/saimamanzoor651/urdu-ocr-codesaviours-si26-saima/blob/main/SI26_Week4_Saima.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Week 4
## Fine-Tune TrOCR on Urdu Datase
##  Overview

In this week, we fine-tune the Microsoft TrOCR model on our Urdu OCR dataset. The model is trained using the custom dataset prepared in Week 3, evaluated on the test set, and finally saved to Google Drive for future use.


In [1]:
!pip install -q \
transformers==4.41.2 \
tokenizers==0.19.1 \
sentencepiece \
protobuf \
tiktoken \
torch \
torchvision \
pandas \
pillow

In [2]:
import os
import torch
import pandas as pd

from PIL import Image
from torch.utils.data import Dataset, DataLoader

from transformers import (
    TrOCRProcessor,
    VisionEncoderDecoderModel,
    AdamW
)

print("Libraries loaded successfully!")

Libraries loaded successfully!


In [3]:
class UrduOCRDataset(Dataset):

    def __init__(self, csv_path, processor):
        self.data = pd.read_csv(csv_path)
        self.processor = processor
        self.base_path = "/content/urdu-ocr-codesaviours-si26-saima"

        print(f"Dataset loaded: {len(self.data)} samples")

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):

        row = self.data.iloc[idx]

        image_path = os.path.join(self.base_path, row["image"])

        image = Image.open(image_path).convert("RGB")

        pixel_values = self.processor(
            image,
            return_tensors="pt"
        ).pixel_values.squeeze()

        labels = self.processor.tokenizer(
            row["text"],
            padding="max_length",
            max_length=128,
            truncation=True,
            return_tensors="pt"
        ).input_ids.squeeze()

        return {
            "pixel_values": pixel_values,
            "labels": labels
        }

In [4]:
import os
print(os.listdir("/content"))

['.config', 'urdu-ocr-codesaviours-si26-saima', 'sample_data']


In [5]:
!git clone https://github.com/saimamanzoor651/urdu-ocr-codesaviours-si26-saima.git

fatal: destination path 'urdu-ocr-codesaviours-si26-saima' already exists and is not an empty directory.


In [6]:
import os

print(os.listdir("/content"))

['.config', 'urdu-ocr-codesaviours-si26-saima', 'sample_data']


In [14]:
from transformers import TrOCRProcessor, VisionEncoderDecoderModel

device = "cuda" if torch.cuda.is_available() else "cpu"

processor = TrOCRProcessor.from_pretrained(
    "microsoft/trocr-base-printed",
    use_fast=False
)

model = VisionEncoderDecoderModel.from_pretrained(
    "microsoft/trocr-base-printed"
)

model.to(device)

model.config.decoder_start_token_id = processor.tokenizer.cls_token_id
model.config.pad_token_id = processor.tokenizer.pad_token_id
model.config.eos_token_id = processor.tokenizer.sep_token_id
model.config.vocab_size = model.config.decoder.vocab_size

print("Model loaded successfully!")

Some weights of VisionEncoderDecoderModel were not initialized from the model checkpoint at microsoft/trocr-base-printed and are newly initialized: ['encoder.pooler.dense.bias', 'encoder.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Model loaded successfully!


## Step 1 – Load the Pretrained TrOCR Model

In this step, we load the pretrained Microsoft TrOCR model and its processor. We also detect whether a GPU is available and move the model to the selected device for faster training.

## Step 2 – Set Up Training

In this step, we prepare the Urdu OCR dataset, split it into training and testing sets, create DataLoaders, and initialize the AdamW optimizer for model training.

In [8]:
processor = TrOCRProcessor.from_pretrained(
    "microsoft/trocr-base-printed",
    use_fast=False
)

dataset = UrduOCRDataset(
    "/content/urdu-ocr-codesaviours-si26-saima/labels.csv",
    processor
)

sample = dataset[0]

print("Sample pixel_values shape:", sample["pixel_values"].shape)
print("Sample labels shape:", sample["labels"].shape)

train_size = int(0.8 * len(dataset))
test_size = len(dataset) - train_size

train_dataset, test_dataset = torch.utils.data.random_split(
    dataset,
    [train_size, test_size]
)

print("Training samples:", train_size)
print("Testing samples:", test_size)

Dataset loaded: 203 samples
Sample pixel_values shape: torch.Size([3, 384, 384])
Sample labels shape: torch.Size([128])
Training samples: 162
Testing samples: 41


In [9]:
from torch.utils.data import DataLoader
from transformers import AdamW

train_loader = DataLoader(
    train_dataset,
    batch_size=4,
    shuffle=True
)

test_loader = DataLoader(
    test_dataset,
    batch_size=4
)

optimizer = AdamW(
    model.parameters(),
    lr=5e-5
)

print("Training batches:", len(train_loader))
print("Testing batches:", len(test_loader))
print("Ready to train!")

/usr/local/lib/python3.12/dist-packages/transformers/optimization.py:588: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


Training batches: 41
Testing batches: 11
Ready to train!


## Step 3 – Train the TrOCR Model

In this step, the TrOCR model is fine-tuned on the Urdu OCR dataset for three epochs. During training, the loss value is monitored to measure the learning progress.

In [10]:
num_epochs = 15

for epoch in range(num_epochs):

    model.train()
    total_loss = 0

    print(f"\nEpoch {epoch+1}/{num_epochs}")
    print("-"*30)

    for batch_idx, batch in enumerate(train_loader):

        pixel_values = batch["pixel_values"].to(device)
        labels = batch["labels"].to(device)

        outputs = model(
            pixel_values=pixel_values,
            labels=labels
        )

        loss = outputs.loss

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        total_loss += loss.item()

        if batch_idx % 10 == 0:
            print(
                f"Batch {batch_idx}/{len(train_loader)} | Loss: {loss.item():.4f}"
            )

    avg_loss = total_loss / len(train_loader)

    print(f"Epoch {epoch+1} Average Loss: {avg_loss:.4f}")

print("\nTraining Complete!")


Epoch 1/15
------------------------------
Batch 0/41 | Loss: 17.3410
Batch 10/41 | Loss: 4.3475
Batch 20/41 | Loss: 3.0942
Batch 30/41 | Loss: 2.8637
Batch 40/41 | Loss: 3.6614
Epoch 1 Average Loss: 4.5779

Epoch 2/15
------------------------------
Batch 0/41 | Loss: 3.1295
Batch 10/41 | Loss: 3.1518
Batch 20/41 | Loss: 3.3612
Batch 30/41 | Loss: 3.4360
Batch 40/41 | Loss: 3.2117
Epoch 2 Average Loss: 3.0458

Epoch 3/15
------------------------------
Batch 0/41 | Loss: 2.6459
Batch 10/41 | Loss: 3.1438
Batch 20/41 | Loss: 3.2836
Batch 30/41 | Loss: 2.5669
Batch 40/41 | Loss: 1.9994
Epoch 3 Average Loss: 2.9702

Epoch 4/15
------------------------------
Batch 0/41 | Loss: 2.4202
Batch 10/41 | Loss: 3.1853
Batch 20/41 | Loss: 2.6636
Batch 30/41 | Loss: 3.4969
Batch 40/41 | Loss: 3.6086
Epoch 4 Average Loss: 3.0136

Epoch 5/15
------------------------------
Batch 0/41 | Loss: 3.0499
Batch 10/41 | Loss: 3.0909
Batch 20/41 | Loss: 2.3847
Batch 30/41 | Loss: 2.7311
Batch 40/41 | Loss: 2.602

## Step 4 – Evaluate the Model

After training, the model is evaluated using the test dataset. The predicted text is compared with the actual labels, and the overall accuracy is calculated.

In [11]:
model.eval()

print("=== Model Evaluation ===")
print()

correct = 0
total = 0

with torch.no_grad():

    for batch in test_loader:

        pixel_values = batch["pixel_values"].to(device)
        labels = batch["labels"]

        generated_ids = model.generate(
            pixel_values,
            max_new_tokens=128,
            num_beams=4,
            early_stopping=True
        )

        # DEBUG: Check what model is actually generating
        print("Generated IDs:", generated_ids)
        print("Raw Output:", processor.batch_decode(generated_ids, skip_special_tokens=False))

        generated_text = processor.batch_decode(
            generated_ids,
            skip_special_tokens=True
        )

        actual_text = processor.batch_decode(
            labels,
            skip_special_tokens=True
        )

        for pred, actual in zip(generated_text, actual_text):

            total += 1

            if pred.strip() == actual.strip():
                correct += 1

            print("Predicted:", pred)
            print("Actual   :", actual)
            print("-" * 50)

accuracy = (correct / total) * 100 if total > 0 else 0

print(f"\nAccuracy: {accuracy:.1f}% ({correct}/{total})")

=== Model Evaluation ===

Generated IDs: tensor([[2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1],
        [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1],
        [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 

## Step 5 – Save the Trained Model

Finally, the trained TrOCR model and processor are saved to Google Drive so they can be reused in future weeks without retraining.

In [12]:
from google.colab import drive

drive.mount('/content/drive')

save_path = "/content/drive/MyDrive/SI26-urdu-ocr-model"

model.save_pretrained(save_path)
processor.save_pretrained(save_path)

print("Model saved successfully!")
print(save_path)

Mounted at /content/drive
Model saved successfully!
/content/drive/MyDrive/SI26-urdu-ocr-model


In [13]:
import pandas as pd

df = pd.read_csv("/content/urdu-ocr-codesaviours-si26-saima/labels.csv")

print(df.head())

                         image  \
0  data/raw/books/book (1).png   
1  data/raw/books/book (2).png   
2  data/raw/books/book (3).png   
3  data/raw/books/book (4).png   
4  data/raw/books/book (5).png   

                                                text  
0  کیا تھا کوچ جو دل سے خوشی کی فوجوں نے لگائے تھ...  
1  غم و الم نے جگر میں لگا دی تھی ایک آگ بنا ہوا ...  
2  زندگی غم نے پریشاں کیا ہوا تھا مجھے یہ فکر کھا...  
3  ہر گھڑی اے دل ہمیں اشکوں کا دریا چاہیے داستان ...  
4  آج ہم حال دل درد آشنا کہنے کو ہیں اس بھری محفل...  


## Results

- Dataset Images: 203
- Training Samples: 162
- Testing Samples: 41
- Final Accuracy: 0
- Training Loss: 0.5693 → 0.0000
- Model Saved Successfully to Google Drive